### ISSUE TITLE
Allow NFTs to include custom data

### ISSUE THREAD

**Issue by ethanfrey** [CosmWasm/cw-plus#440]
A first step on #391

A few people have tried to build on cw721-base but want to add extra metadata to the contract and cannot extend it. Basically, we would need:

```rust
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
pub struct TokenInfo<T> {
    /// The owner of the newly minted NFT
    pub owner: Addr,
    /// Approvals are stored here, as we clear them all upon transfer and cannot accumulate much
    pub approvals: Vec<Approval>,

    /// Identifies the asset to which this NFT represents
    pub name: String,
    /// Describes the asset to which this NFT represents
    pub description: String,
    /// A URI pointing to an image representing the asset
    pub image: Option<String>,

    pub extension: Option<T>,
}
```
([source](https://github.com/CosmWasm/cw-plus/blob/main/contracts/cw721-base/src/state.rs#L9-L21))

The default implementation can simply leave `extension: None`. (Or use `extension: Empty`, removing the Option).

This can be done and then we make all query and execute methods accept this T. However, when thinking about it, it would be a great first example of using a struct to represent a contract, as described in #391

**Comment by hashedone** [comment]
I am not sure if it is worth it for now waiting for #391 - I am doing some things around this after hours, but for now I don't put time on this unless we are agreed on this idea. Adding generic `extension` is relatively small change and it can be done in current environment (I am not sure how important is it).

One thing I would change here - make `extension` just `T` and leave it to the user to decide if it is optional (`T` may always be an option itself), making it `Empty` by default.

**Comment by ethanfrey** [comment]
> One thing I would change here - make extension just T and leave it to the user to decide if it is optional (T may always be an option itself), making it Empty by default.

Sounds good to use T.
Making it `Option<Empty>` by default is nice, so it is compatible with current serialised data that don't have that field.

**Comment by ethanfrey** [comment]
I wouldn't wait until #391 but do some refactoring here to start sketching out what #391 (after code-gen) could look like on a concrete use-case.

**Comment by ethanfrey** [comment]
Closed by #445 

Example in #450
